src/hotspot/share/compiler/compilationPolicy.cpp

```diff
@@ -453,12 +455,20 @@ void CompilationPolicy::initialize() {
 #ifdef COMPILER2
       c2_size = C2Compiler::initial_code_buffer_size();
 #endif
-      size_t buffer_size = c1_only ? c1_size : (c1_size/3 + 2*c2_size/3);
+      size_t buffer_size = 0;
+      if (c1_only) {
+        buffer_size = c1_size;
+      } else if (c2_only) {
+        buffer_size = c2_size;
+      } else {
+        buffer_size = c1_size/3 + 2*c2_size/3;
+      }
       int max_count = (ReservedCodeCacheSize - (CodeCacheMinimumUseSpace DEBUG_ONLY(* 3))) / (int)buffer_size;
       if (count > max_count) {
         // Lower the compiler count such that all buffers fit into the code cache
-        count = MAX2(max_count, c1_only ? 1 : 2);
+        count = MAX2(max_count, min_count);
       }
+      assert((!c1_only && !c2_only) || count <= active_cpus, "Too many threads: %d", count);
```

Comment:
Should it be the general rule: don't create more compiler threads than available cpus?

Reply:
Except when specified on command line with `-XX:CICompilerCount=n`.
Actually your changes does not take this flag into account.

Reply:
I think we should allow users to run with more threads than CPUs, if they really want it. (This is also handy for testing the heuristics for very large CPU counts without having too many real CPUs.) What we surely need to do is to teach _heuristics_ never go over the CPU count. I now added the more precise test that explores `CICompilerCount` settings as well, verifying heuristics works as expected.